From: Vitaliy Gusev <vgusev@openvz.org>
Date: Tue, 14 Oct 2008 15:10:44 +0000 (+0400)
Subject: CPT: revert check on sk_reuse>1
X-Git-Tag: sync-2.6.27-15.10.08~4
X-Git-Url: http://git.openvz.org/?p=linux-2.6.26-openvz;a=commitdiff_plain;h=6d18ba377cfa3e86ee830fe6a5fce52b8fd51039

CPT: revert check on sk_reuse>1

Revert commit ac6f78192054784f02dd47f8e6d7d1c8d75ab173
("[INET]: sk_reuse is valbool", Author: Gerrit Renker <gerrit@erg.abdn.ac.uk>)

Check on sk_reuse>1 is needed as during restore "bind" fails
for sockets that was bound to the same port for same sk_family and also
fails for sockets that are bound to the same port for other sk_family
(Example: sshd that listen to port 22 IPv4 and IPv6 any address).

For additional information see open_listening_sockets() and open_socket().

Related to the bug#1034
http://bugzilla.openvz.org/show_bug.cgi?id=1034

Signed-off-by: Vitaliy Gusev <vgusev@openvz.org>
Signed-off-by: Pavel Emelyanov <xemul@openvz.org>
---

diff --git a/net/ipv4/inet_connection_sock.c b/net/ipv4/inet_connection_sock.c
index 5bfa408..2fbce13 100644
--- a/net/ipv4/inet_connection_sock.c
+++ b/net/ipv4/inet_connection_sock.c
@@ -146,6 +146,8 @@ int inet_csk_get_port(struct sock *sk, unsigned short snum)
 	goto tb_not_found;
 tb_found:
 	if (!hlist_empty(&tb->owners)) {
+		if (sk->sk_reuse > 1)
+			goto success;
 		if (tb->fastreuse > 0 &&
 		    sk->sk_reuse && sk->sk_state != TCP_LISTEN) {
 			goto success;
